Introduce ContiguouslyStored protocol (#956)
apple/swift-evolution: proposals/NNNN-contiguous-collection.md

# Introduce Contiguous Collection Protocols

> * Proposal: [SE-NNNN](NNNN-filename.md)

Reviewer: ```suggestion
* Proposal: [SE-NNNN](NNNN-contiguous-collection.md)
```

  func withUnsafeBufferPointer<R>(
    _ body: (UnsafeBufferPointer<Element>) throws -> R
  ) rethrows -> R
}

/// A collection that supports mutable access to its underlying contiguous
/// storage.
public protocol MutableContiguousCollection: ContiguousCollection, MutableCollection
where SubSequence: MutableContiguousCollection {
  /// Calls the given closure with a pointer to the array's mutable contiguous
  /// storage.
  mutating func withUnsafeMutableBufferPointer<R>(
    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
  ) rethrows -> R
}
```

Conformances will be added for the following types:
- `Array`, `ArraySlice` and `ContiguousArray` will conform to `MutableContiguousCollection`
- `UnsafeBufferPointer` will conform to `ContiguousCollection`
- `UnsafeMutableBufferPointer` will conform to `MutableContiguousCollection`
- `Slice` will conditionally conform:
    - to `ContiguousCollection where Base: ContiguousCollection`
    - to `MutableContiguousCollection where Base: MutableContiguousCollection`

In addition, the following customization point should be added to
`MutableCollection`, with a default implementation when the collection is

> mutably contiugously stored:

Reviewer: ```suggestion
mutably contiguously stored:
```

```swift
protocol MutableCollection {
  /// Call `body(p)`, where `p` is a pointer to the collection's
  /// mutable contiguous storage.  If no such storage exists, it is
  /// first created.  If the collection does not support an internal
  /// representation in a form of mutable contiguous storage, `body` is not
  /// called and `nil` is returned.
  public mutating func withUnsafeMutableBufferPointerIfSupported<R>(
    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
  ) rethrows -> R?
}
```

This customization point already exists with an underscore in the standard
library (it just returns `nil` by default), and should be exposed to
general users, but could have a default implementation when the collection

> `Mutable` Use of this entry point can provide significant speedups in some

Reviewer: but could have a default implementation when the collection **???**

Author: ah right, "is contiguously stored".

algorithms, e.g. our current

> [`sort`](https://github.com/apple/swift/blob/master/stdlib/public/core/Sort.swift#L249)

Reviewer: ```suggestion
[`sort`](https://github.com/apple/swift/blob/6662ccc16dba27418eefd3cb7856bddda5a33386/stdlib/public/core/Sort.swift#L249)
```

which needs to move elements of a collection back and forth between
some storage.

## Source compatibility

> These are additive changes and do not affect ABI stability.

Reviewer: ```suggestion
These are additive changes and do not affect source compatibility.
```

`withUnsafeMutableBufferPointer`, which are only available on the concrete
types. Given the usefulness of these methods, they should also be made
available generically.

## Proposed solution

Introduce two new protocols, with requirements representing the with-unsafe capabilities of `Array` & co:

```swift
/// A collection that supports access to its underlying contiguous storage.
public protocol ContiguousCollection: RandomAccessCollection
where SubSequence: ContiguousCollection {
  /// Calls a closure with a pointer to the array's contiguous storage.
  func withUnsafeBufferPointer<R>(
    _ body: (UnsafeBufferPointer<Element>) throws -> R
  ) rethrows -> R
}

/// A collection that supports mutable access to its underlying contiguous
/// storage.
public protocol MutableContiguousCollection: ContiguousCollection, MutableCollection
where SubSequence: MutableContiguousCollection {
  /// Calls the given closure with a pointer to the array's mutable contiguous
  /// storage.
  mutating func withUnsafeMutableBufferPointer<R>(
    _ body: (inout UnsafeMutableBufferPointer<Element>) throws -> R
  ) rethrows -> R
}
```

> Note, the mutable variant takes the buffer `inout`. This is to facilitate the ergnomics 

Reviewer: ```suggestion
Note, the mutable variant takes the buffer `inout`. This is to facilitate the ergonomics 
```

guaranteed to give you the same pointer with each call. For example, a packed small string
implementation may be giving you a pointer to that string temporarily expanded to a buffer.

Use of this entry point can provide significant speedups in some
algorithms, e.g. our current
[`sort`](https://github.com/apple/swift/blob/6662ccc16dba27418eefd3cb7856bddda5a33386/stdlib/public/core/Sort.swift#L249)
which needs to move elements of a collection back and forth between
some storage.

## Source compatibility

These are additive changes and do not affect source compatibility.

## Effect on ABI stability

These are additive changes and do not affect ABI stability.

## Alternatives considered

Some collections are not fully contiguous, but instead consist of multiple contiguous 
regions (for example, a ring buffer is one or two separate contiguous regions). A protocol that
exposed a collection of contiguous regions could be implemented on top of this protocol.

The `inout` argument to the closure in the mutating variant is debatable. It does imply the user
can change the buffer to a totally different one. Nonetheless, this is better handled in documentation,

> since the improved ergnomics of the `inout` version are considerable. It would also be a source-breaking

Reviewer: ```suggestion
since the improved ergonomics of the `inout` version are considerable. It would also be a source-breaking
```
